x-pack/plugins/apm/public/components/shared/FilterBar/ServiceEnvFilter.tsx

```diff
@@ -0,0 +1,94 @@
+import { EuiFilterGroup, EuiFilterButton, EuiSelect } from '@elastic/eui';
+import React, { useEffect, useRef, useState } from 'react';
+import { useFetcher } from '../../../hooks/useFetcher';
+import { loadServiceEnvironments } from '../../../services/rest/apm/services';
+import { useLocation } from '../../../hooks/useLocation';
+import { useUrlParams } from '../../../hooks/useUrlParams';
+import { history } from '../../../utils/history';
+import { fromQuery, toQuery } from '../Links/url_helpers';
+
+function without<T>(obj: T, key: string): T {
+  const result: { [objKey: string]: unknown } = Object.assign({}, obj);
+  delete result[key];
+  return result as T;
+}
+
+export const ServiceEnvFilter: React.FC = () => {
+  const location = useLocation();
+  const { urlParams } = useUrlParams();
+  const { start, end, kuery, environment } = urlParams;
+  const { data = [] } = useFetcher(
+    () => loadServiceEnvironments({ start, end, kuery, environment: '' }),
+    [start, end, kuery]
+  );
+  const availableEnvironments = ['Any', ...data, 'Not set'];
+  const [selectedValue, setSelectedValue] = useState(availableEnvironments[0]);
+  const isPostEffect = useRef(false);
+
+  function updateEnvironmentUrl(environment?: string) {
+    if (!environment || environment === availableEnvironments[0]) {
+      history.push({
+        ...location,
+        search: fromQuery(without(toQuery(location.search), 'environment'))
+      });
+    } else if (
+      environment === availableEnvironments[availableEnvironments.length - 1]
+    ) {
+      history.push({
+        ...location,
+        search: fromQuery({
+          ...toQuery(location.search),
+          environment: 'Null'
+        })
+      });
+    } else {
+      history.push({
+        ...location,
+        search: fromQuery({
+          ...toQuery(location.search),
+          environment
+        })
+      });
+    }
+  }
```

Comment:
You should be able to simplify this to a single `history.push`. `EuiSelect` takes both a `text` and a `value` field. If you use the value field, you shouldn't need the duplication here.